Vignettes (#144)
r-spatial/qgisprocess: vignettes/qgisprocess.Rmd

---
title: "Basic usage of qgisprocess"
author: "Jannes Muenchow & Floris Vanderhaeghe"
date: "Last updated: yyyy-mm-dd"
output:
  html_document: 
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: no
    toc_depth: 2
vignette: >
  %\VignetteIndexEntry{Basic usage of qgisprocess}
  %\VignetteEncoding{UTF-8}

>   %\VignetteEngine{knitr::knitr}

Reviewer: ```suggestion
  %\VignetteEngine{knitr::rmarkdown}
```

This appears to be the cause of most failing checks. Beware you also need to apply https://github.com/r-spatial/qgisprocess/pull/144#issuecomment-1483160720 for systems that lack QGIS (included in the checks).

---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Setting up the system

**qgisprocess** is basically a wrapper around the standalone command line tool [`qgis_process`](https://github.com/qgis/QGIS-Documentation/issues/5803).
Therefore, you need to have installed QGIS on your system as well as third-party providers such as GRASS and SAGA to access and run all geoalgorithms provided through `qgis_process` from within R.

To facilitate using **qgisprocess**, we have created a docker image that already comes with the needed software packages.
You can pull it from Github's container registry by running:

```{sh, eval=FALSE}
docker pull ghcr.io/geocompx/docker:qgis
```

For a more detailed introduction on how to get started with docker, please refer to https://github.com/geocompx/docker.

# Package configuration

Since **qgisprocess** depends on the command line tool `qgis_process`, it already tries to detect `qgis_process` on your system when it is being loaded, and complains if it cannot find it.

```{r}
library("qgisprocess")
```

> In fact, once **qgisprocess** has detected `qgis_process`, it will cache among others the path to the executable, the QGIS version and the list of known algorithms.

Reviewer: ```suggestion
When loading **qgisprocess** for the first time, it will cache among others the path to `qgis_process`, the QGIS version and the list of known algorithms.
When loading **qgisprocess** in later R sessions, the cache file is read instead for speed-up, on condition that it is still valid.
```

```{r algs}
algs <- qgis_algorithms()
algs
```

Since we have also installed GRASS and SAGA, nearly 1000 geoalgorithms are at our disposal.
To find out about a specific geoalgorithm and a description of its arguments, use `qgis_show_help()`, e.g.:

```{r help, eval=FALSE}
qgis_show_help("native:buffer")
## Buffer (native:buffer)
## 
## ----------------
## Description
## ----------------
## This algorithm computes a buffer area for all the features in an input layer, using a fixed or dynamic distance.
## 
## The segments parameter controls the number of line segments to use to approximate a quarter circle when creating rounded offsets.
## 
##...
```

To find out the arguments of a specific geoalgorithm, run:

```{r args-buffer}
qgis_arguments("native:buffer")
```

And finally run it with `qgis_run_algorithm()`:

```{r buffer}
# if needed, first install spDataLarge:
# remotes::install_github("Nowosad/spDataLarge")
data("random_points", package = "spDataLarge")
out <- qgis_run_algorithm("native:buffer", INPUT = random_points, DISTANCE = 50)
```

As a convenience to the user, `qgis_run_algorithm()` reports all unspecified and automatically chosen arguments.

> If you want to have even more information on what is going on in the background, set `.quiet` to `FALSE`.

Author: ok, I rewrote "additionally set argument ..."

```{r args}
qgis_arguments("grass7:r.slope.aspect")
```

`qgis_outputs()` shows the different outputs that will be calculated:

```{r outputs}
qgis_outputs("grass7:r.slope.aspect")
```

Now let us calculate the terrain attributes.

```{r slopeaspect, message=FALSE}
library("terra")
# attach digital elevation model from Mt. Mongón (Peru)
dem <- rast(system.file("raster/dem.tif", package = "spDataLarge"))

info <- qgis_run_algorithm(alg = "grass7:r.slope.aspect", elevation = dem)
                           
```

Just printing the `info` object shows which output objects have been made:

```{r info}
info
```

Combine these output rasters as a multi-layered `SpatRaster` object and plot it:

```{r combine1, message=FALSE}
# just keep the names of output rasters
nms <- qgis_outputs("grass7:r.slope.aspect")$name
# read in the output rasters 
r <- info[nms] |>
  unlist() |>
  terra::rast() |>
  as.numeric()
names(r) <- nms
# plot the output
plot(r)
```

An alternative way to combine the rasters is given below.

```{r combine2, message=FALSE}
r <- lapply(info[nms], \(x) as.numeric(qgis_as_terra(x))) |> 
  rast()
plot(r)
```

> ## Piping

Reviewer: ```suggestion
# Piping
```

Perhaps consider outside 'Basic usage'?

`qgis_process` does not lend itself naturally to piping because its first argument is the name of a geoalgorithm instead of a data object.
`qgis_pipe()` circumvents this by accepting a `.data` object as its first argument, and pipes this data object into the the first argument of a geoalgorithm assuming that the specified geoalgorithm in fact needs a data input object as its first argument.

```{r buf-pipe}
system.file("longlake/longlake_depth.gpkg", package = "qgisprocess") |>
  qgis_pipe("native:buffer", DISTANCE = 50)
```

> In case of a `qgis_result` object, `qgis_pipe()` automatically tries to select an element named `OUTPUT`.

Reviewer: ```suggestion
If `.data` is a `qgis_result` object, `qgis_pipe()` automatically tries to select an element named `OUTPUT`.
```

However, if the output has another name (e.g., `DEM_PREPROC` as in the example below) or if there are multiple output elements to choose from (e.g., `sagang:sagawetnessindex` has four output rasters, check with `qgis_outputs("sagang:sagawetnessindex")`, you can specify the wanted output object via the `.select` argument.

```{r qgis-pipe}
dem <- system.file("raster/dem.tif", package = "spDataLarge")
# in case you need to enable the SAGA next generation algorithms, run the following lines:
# qgis_plugins()
# qgis_enable_plugins("processing_saga_nextgen")
qgis_run_algorithm(algorithm = "sagang:sinkremoval", DEM = dem, 
                   METHOD = 1) |>
  qgis_pipe("sagang:sagawetnessindex", .select = "DEM_PREPROC")
```

When piping, `qgis_pipe()` automatically cleans up after you by deleting intermediate results.
This avoids cluttering your system when running geoalgorithms on large spatial data files.
To turn of this behavior, set `.clean` to `FALSE`.

> Of course, you can also use the native pipe.

Reviewer: ```suggestion
Of course, you can also pipe to `qgis_run_algorithm()` by manually extracting the `OUTPUT` object and redirecting it to the appropriate input argument of the next processing step.
```

The choice here is not about using (native) pipes or not, as pipes are used with `qgis_pipe()` too. Also 'native' vs **magrittr**'s pipe is not relevant for the point made, except for syntax difference if using the placeholder.

This avoids ambiguity and allows for greater flexibility though it might not be as convenient as `qgis_pipe()`.
For example, intermediate results remain on disk for the duration of your R session.

```{r native pipe}
out = qgis_run_algorithm(algorithm = "sagang:sinkremoval", DEM = dem, 
                         METHOD = 1) |>

>    (\(x) qgis_run_algorithm(algorithm = "sagang:sagawetnessindex",

Reviewer: FWIW I doubt whether the lambda syntax as shorthand for anonymous functions is well known among users (it's only there since R 4.1), but of course it's valid, but it may distract from the point you are making. (The placeholder syntax is simpler)

Author: yes, I plan to use the placeholder as soon as R 4.3 is available on Ubuntu 22.04

Reviewer: 4.3.0 is available, if one follows the R installation instructions for Ubuntu at https://cran.r-project.org/bin/linux/ubuntu/. The regular Ubuntu repo does not follow R versions this quickly, if at all.

```bash
$ apt policy r-base
r-base:
  Installed: 4.3.0-1.2204.0
  Candidate: 4.3.0-1.2204.0
  Version table:
 *** 4.3.0-1.2204.0 500
        500 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ Packages
        100 /var/lib/dpkg/status
     4.2.3-1.2204.0 500
        500 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ Packages
     4.2.2.20221110-1.2204.0 500
        500 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ Packages
     4.2.2-1.2204.0 500
        500 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ Packages
     4.2.1-3.2204.0 500
        500 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ Packages
     4.2.1-2.2204.0 500
        500 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ Packages
     4.2.1-1.2204.0 500
        500 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ Packages
     4.2.0-1.2204.0 500
        500 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ Packages
     4.1.2-1ubuntu2 500
        500 http://ftp.belnet.be/ubuntu jammy/universe amd64 Packages
        500 http://ftp.belnet.be/ubuntu jammy/universe i386 Packages
```

Author: sorry, I have been unclear. Of course, you are right, that we would have to wait forever for the latest R version using the regular Ubuntu repo. I was referring to the [rocker/geospatial Dockerfiles](https://github.com/rocker-org/rocker-versioned2/pkgs/container/geospatial/versions) on which the ghcr.io/geocompx/docker:qgis image is based. But in fact, the geospatial images already come with R >4.3. When will this come into effect in the geocompx images, @Robinlovelace?
